sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/testing/InMemoryMetricReader.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.testing;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import io.opentelemetry.sdk.metrics.export.MetricProducer;
+import io.opentelemetry.sdk.metrics.export.MetricReader;
+import io.opentelemetry.sdk.metrics.export.MetricReaderFactory;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * A {@link MetricReader} implementation that can be used to test OpenTelemetry integration.
+ *
+ * <p>Can be created using {@code InMemoryMetricReader.create(sdkMeterProvider)}
+ *
+ * <p>Example usage:
+ *
+ * <pre><code>
+ * public class InMemoryMetricReaderExample {
+ *   private final InMemoryMetricReader reader = new InMemoryMetricReader();
+ *   private final SdkMeterProvider sdkMeterProvider = SdkMeterProvider.builder().registerMetricReader(reader).build();
+ *   private final Meter meter = sdkMeterProvider.get("example");
+ *   private final LongCounter metricCallCount = meter.counterBuilder("num_collects");
+ *
+ *   public void printMetrics() {
+ *     metricCallCount.add(1);
+ *     System.out.println(reader.collectAllMetrics());
+ *   }
+ *
+ *   public static void main(String[] args) {
+ *     InMemoryMetricReaderExample example = new InMemoryMetricReaderExample();
+ *     example.printMetrics();
+ *   }
+ * }
+ * </code></pre>
+ */
+public class InMemoryMetricReader implements MetricReader, MetricReaderFactory {
+  // Note: we expect the `apply` method of `MetricReaderFactory` to be called
```

Comment:
Even for testing think it's worth sticking to the same pattern we did to Prometheus (here it really should just be returning `::new` :)

Reply:
We can't, you loose access to the instantiated class if you do so, which means you never have access to collect metrics.

This is designed so you can directly call "collect" and extract metrics.  (also, should this class be in the metric-testing artifact?)